source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,624 @@
+    if( az_result_failed( xCoreResult = az_iot_hub_client_methods_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                                        xTopicSpan, &xOutEmbeddedRequest ) ) )
+    {
+        AZLogWarn( ( "Direct method topic parsing failed: 0x%08x", xCoreResult ) );
+        xResult = eAzureIoTHubClientTopicNoMatch;
+    }
+    else
+    {
+        /* Verify the received publish is for the feature we have subscribed to. */
+        AZLogDebug( ( "Direct method topic: %.*s  with method payload : %.*s.",
+                      xMQTTPublishInfo->usTopicNameLength,
+                      xMQTTPublishInfo->pcTopicName,
+                      xMQTTPublishInfo->xPayloadLength,
+                      xMQTTPublishInfo->pvPayload ) );
+
+        if( pxContext->_internal.callbacks.xMethodCallback )
+        {
+            xMethodRequest.pvMessagePayload = xMQTTPublishInfo->pvPayload;
+            xMethodRequest.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->xPayloadLength;
+            xMethodRequest.pucMethodName = az_span_ptr( xOutEmbeddedRequest.name );
+            xMethodRequest.usMethodNameLength = ( uint16_t ) az_span_size( xOutEmbeddedRequest.name );
+            xMethodRequest.pucRequestID = az_span_ptr( xOutEmbeddedRequest.request_id );
+            xMethodRequest.usRequestIDLength = ( uint16_t ) az_span_size( xOutEmbeddedRequest.request_id );
+
+            AZLogDebug( ( "Invoking Methods Callback." ) );
+            pxContext->_internal.callbacks.xMethodCallback( &xMethodRequest, pxContext->_internal.pvCallbackContext );
+            AZLogDebug( ( "Returned from Methods Callback." ) );
+
+            xResult = eAzureIoTHubClientSuccess;
+        }
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming device twin messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientDeviceTwinProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                       void * pxPublishInfo )
+{
+    AzureIoTHubClientResult_t xResult;
+    AzureIoTHubClientTwinResponse_t xTwinResponse = { 0 };
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
+    az_result xCoreResult;
+    az_iot_hub_client_twin_response xOutRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pcTopicName, xMQTTPublishInfo->usTopicNameLength );
+    uint32_t ulRequestID = 0;
+
+    /* Failed means no topic match */
+    xCoreResult = az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                               xTopicSpan, &xOutRequest );
+
+    if( az_result_failed( xCoreResult ) )
+    {
```

Comment:
looks much cleaner .... can we make all the process symmetric like this .

core = parse_api();

if (failed( core))
{
}
else
{
}